### PR TITLE
Tracing for C++

### PR DESCRIPTION
**Pull request by rherrmannr**

### CODE REVIEW

**Comment by rherrmannr** [comment]
fix #2437 
fix https://github.com/Yakindu/sctpro/issues/1823

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CGenerator.xtend [hunk 1]
```diff
@@ -65,7 +65,7 @@ class CGenerator implements IExecutionFlowGenerator {
 		locations.configure(flow.typesModule.h, entry.libraryOutput, types, getSkipLibraryFiles(entry))
 		locations.configure(flow.module.h, entry.headerOutput, statemachineHeader)
 		locations.configure(flow.module.c, entry.sourceOutput, statemachineSource)
-		if (flow.timed || !flow.operations.empty || entry.tracingEnterState || entry.tracingExitState) {
+		if (flow.timed || !flow.operations.empty || entry.tracingUsed) {
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CGenerator.xtend]: 👍

File: plugins/org.yakindu.sct.generator.cpp/library/FeatureTypeLibrary.xmi [hunk 2]
```diff
@@ -49,4 +49,17 @@
...
+  <types name="Tracing"
+      comment="Enables generation of tracing callbacks."
+      optional="true">
+    <parameters
+        name="enterState"
+        comment="Whether to generate a callback that is used when a state is entered."
+        optional="true"
+        parameterType="BOOLEAN"/>
+    <parameters
+        name="exitState"
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.cpp/library/FeatureTypeLibrary.xmi]: both optional maybe?

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CppCodeGeneratorModule.java [hunk 3]
```diff
@@ -61,11 +65,22 @@ public void configure(GeneratorEntry entry, Binder binder) {
...
 	}
+	
+	protected void bindTracingProperty(GeneratorEntry entry, Binder binder) {
+		FeatureParameterValue traceEnterFeature = entry.getFeatureParameterValue(FEATURE_TRACING,
+				PARAMETER_TRACING_ENTER_STATE);
+		FeatureParameterValue traceExitFeature = entry.getFeatureParameterValue(FEATURE_TRACING,
+				PARAMETER_TRACING_EXIT_STATE);
+		boolean traceEnter = traceEnterFeature != null ? traceEnterFeature.getBooleanValue() : false;
+		boolean traceExit = traceExitFeature != null ? traceEnterFeature.getBooleanValue() : false;
+		binder.bind(Boolean.class).annotatedWith(Names.named(ADD_TRACES)).toInstance(traceEnter || traceExit);
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CppCodeGeneratorModule.java]: Why is this needed additionally to `GenModelEntries.tracingUsed()`?

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/StatemachineHeader.xtend [hunk 4]
```diff
@@ -170,6 +176,22 @@ class StatemachineHeader extends org.yakindu.sct.generator.c.files.StatemachineH
...
+	def createTracingScope(ExecutionFlow it) {
+		'''
+		class «tracingModule»
+		{
+			public:
+			virtual ~«tracingModule»() = 0;
+			
+			«IF entry.tracingEnterState»virtual void «enterStateTracingFctID»(«module»::«statesEnumType» state) = 0;«ENDIF»
+			
+			«IF entry.tracingExitState»virtual void «exitStateTracingFctID»(«module»::«statesEnumType» state) = 0;«ENDIF»
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/StatemachineHeader.xtend]: Maybe an empty implementation that does nothing instead of pure virtual would be a good idea for all functions, this would allow the user to just provide a new instance of this class before he implements his own tracing class.

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/StatemachineHeader.xtend [hunk 5]
```diff
@@ -234,6 +256,10 @@ class StatemachineHeader extends org.yakindu.sct.generator.c.files.StatemachineH
 		return ''''''
 	}
 
+	def createInlineTR_Destructor(ExecutionFlow it) {
+		'''inline «module»::«tracingModule»::~«tracingModule»() {}'''
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/StatemachineHeader.xtend]: You could just provide the empty implementation where you declared the destructor above.

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CppCodeGeneratorModule.java (hunk 3 above)
Comment (rherrmannr) [reply plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CppCodeGeneratorModule.java]: I'm sorry. Now I got what you mean.

I could inject GenmodelEntries ``@Inject GenmodelEntries entries;``

and replace the whole thing with 
``binder.bind(Boolean.class).annotatedWith(Names.named(ADD_TRACES)).toInstance(entries.tracingUsed(entry));``

Comment (rherrmannr) [reply plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CppCodeGeneratorModule.java]: And I cannot use ``GenModelEntries.tracingUsed(entry)``, because the functions are not static

**Comment by rherrmannr** [comment]
fix #564
